### PR TITLE
feat(rust): support `define` functionality

### PR DESCRIPTION
**Pull request by hyf0**
### Description

- https://esbuild.github.io/api/#define

### CODE REVIEW

**Comment by hyf0** [comment]
* **#1884** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/1884?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1883** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/1883?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1882** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/1882?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1881** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/1881?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1880** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/1880?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @hyf0 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/rolldown/src/utils/pre_process_ecma_ast.rs [hunk 1]
```diff
@@ -46,10 +50,21 @@ pub fn pre_process_ecma_ast(
...
+  ast.program.with_mut(|fields| -> anyhow::Result<()> {
+    let define_config =
+      ReplaceGlobalDefinesConfig::new(&bundle_options.define).map_err(|errs| {
+        anyhow::format_err!(
+          "Failed to generate defines config from {:?}. Got {:#?}",
+          bundle_options.define,
+          errs
+        )
+      })?;
+    ReplaceGlobalDefines::new(fields.allocator, define_config).build(fields.program);
```
Comment (IWANABETHATGUY) [inline crates/rolldown/src/utils/pre_process_ecma_ast.rs:62]: Construct the `ReplaceGlobalConfig` looks none trivial https://github.com/oxc-project/oxc/blob/910f2a0a0a7a4298923afc88511d089843d02d3a/crates/oxc_minifier/src/ast_passes/replace_global_defines.rs#L30-L54, do we need to share them with `Arc`?

Comment (hyf0) [reply crates/rolldown/src/utils/pre_process_ecma_ast.rs:62]: Good catch. @Boshen cc

Comment (hyf0) [reply crates/rolldown/src/utils/pre_process_ecma_ast.rs:62]: Guess this also requires https://github.com/oxc-project/oxc/issues/4284.

Comment (Boshen) [reply crates/rolldown/src/utils/pre_process_ecma_ast.rs:62]: https://github.com/oxc-project/oxc/blob/69dc6c9cb1f93209e5a0bc003918d1bc6f50e455/crates/oxc_minifier/src/plugins/replace_global_defines.rs#L17 it's `Arc`ed.

Comment (IWANABETHATGUY) [reply crates/rolldown/src/utils/pre_process_ecma_ast.rs:62]: Yeah, the `Defiens` is `Arc`ed in oxc side,but we still create the `DefineConfig` per file.

**Pull request merged**